Fixes part of #193

vdeenadh opened the pull request:
VRG now first protects PV data (via VR) and then protects PV cluster data (via S3 upload).

DRPC still needs to check if Primary VRG is healthy before taking action on relocate; this is not included in this PR.

BenamarMk reviewed (commented):
Veera, I added minor comments.  However, I would like to propose a suggestion.  Refactoring is good, but refactoring with renaming is very difficult to follow.  Moreover, refactoring is good to do if and only if you have a solid unit test to back it up.  Refactoring with not complete uni test is asking for lot of trouble.  In this PR, you renamed many functions, and also you moved code around.  I couldn't follow the change easily.  Even then, I am not sure if part of it is missed.  As an example, you refactored this function `protectPVCsAsPrimary` and removed a portion of code from `protectPVC`, and that portion of code returns `requeue` and `!skip` which I didn't see handled in `protectPVCsAsPrimary`.  So what I am saying is that; it would have been easier if this PR has been split into two.  One for the renaming, and another one for the actual code change.  If the renaming was kept to a minimal, then that wouldn't be a problem.  Anyway, i am not suggesting to split it now, just in the future maybe we follow this strategy.

vdeenadh commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -1141,10 +1139,18 @@ func (v *VRGInstance) undoPVRetentionForPVC(pvc corev1.PersistentVolumeClaim, lo
 
-// Upload PV to the list of S3 stores in the VRG spec
-func (v *VRGInstance) uploadPVToS3Profiles(pvc corev1.PersistentVolumeClaim) (err error) {
+// protectPVClusterDatauploads the PV related cluster data of the given
```
Ack.

vdeenadh commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -773,16 +775,29 @@ func (v *VRGInstance) reconcileVRsAsPrimary() bool {
+			continue
+		}
 
+		// Protect the PV cluster data (in etcd)
```
All k8s resources or objects (in etcd) are collectively called as Cluster Data per definition. In addition, each k8s object has a metadata struct embedded in it and can be confusing.  Most importantly, I think we should start using the official term, which is "[Cluster Data](https://kubernetes.io/docs/reference/glossary/?fundamental=true&architecture=true#term-etcd)" that is defined under etcd.

vdeenadh replied:
PV resource would work for initial release, but in future, we expect to also include PV snapshot resources. Another term is [Kubernetes objects](https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/): would the term `PV Kubernetes Objects` include both PV and PV snapshot objects?

vdeenadh commented:
@BenamarMk Did you notice that there are two commits: one has just renames with no functional changes?

> Veera, I added minor comments. However, I would like to propose a suggestion. Refactoring is good, but refactoring with renaming is very difficult to follow. Moreover, refactoring is good to do if and only if you have a solid unit test to back it up. Refactoring with not complete uni test is asking for lot of trouble. In this PR, you renamed many functions, and also you moved code around. I couldn't follow the change easily. Even then, I am not sure if part of it is missed. As an example, you refactored this function `protectPVCsAsPrimary` and removed a portion of code from `protectPVC`, and that portion of code returns `requeue` and `!skip` which I didn't see handled in `protectPVCsAsPrimary`. So what I am saying is that; it would have been easier if this PR has been split into two. One for the renaming, and another one for the actual code change. If the renaming was kept to a minimal, then that wouldn't be a problem. Anyway, i am not suggesting to split it now, just in the future maybe we follow this strategy.

vdeenadh commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -1823,14 +1832,21 @@ func (v *VRGInstance) deleteVR(vrNamespacedName types.NamespacedName, log logr.L
 	pvc.ObjectMeta.Annotations[pvcVRAnnotationProtectedKey] = pvcVRAnnotationProtectedValue
 
 	if err := v.reconciler.Update(v.ctx, pvc); err != nil {
+		msg := "Failed to add protected annotatation to PVC"
```
Good suggestion towards consistency.  Applies to many places in the code.  Created #221.

BenamarMk commented:
@vdeenadh I didn't notice that they were to commits.  Great.  My bad.

ShyamsundarR commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -951,20 +965,26 @@ func (v *VRGInstance) preparePVCForVRProtection(pvc *corev1.PersistentVolumeClai
 }
 
-func (v *VRGInstance) protectPVC(pvc *corev1.PersistentVolumeClaim,
+func (v *VRGInstance) prepareToProtectPVC(pvc *corev1.PersistentVolumeClaim,
```
If above becomes `preparePVCForReplication` this possibly becomes `preparePVCForVRReplication`

vdeenadh replied:
As I mentioned earlier, with the changes I made, I don't know see why this method applies only to VR.

ShyamsundarR replied:
Currently we are VR specific, it is easier to reason on the present. When we get to the future we can look at renaming is more appropriately?

ShyamsundarR commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -1141,10 +1139,18 @@ func (v *VRGInstance) undoPVRetentionForPVC(pvc corev1.PersistentVolumeClaim, lo
 	for _, s3ProfileName := range v.instance.Spec.S3ProfileList {
 		if err := v.uploadPV(s3ProfileName, pvc); err != nil {
+			msg := "Failed to upload PV cluster data to s3Profile " + s3ProfileName
+			v.updateProtectedPVCCondition(pvc.Name, PVCError, msg)
```
We need to split this into 2 conditions. I know this is prep PR, but we should avid fiddling conditions in here, and handle it at a higher layer (afaict).

vdeenadh replied:
I don't believe you're suggesting to add the condition in the caller, `protectPVCsAsPrimary` that calls three other methods, each of which add conditions inside them. Hence, I thought  it will be consistent to add conditions inside this method.

ShyamsundarR replied:
Are you starting with a condition per PVC or a global VRG condition (I would assume the latter to get the feature going), if the former then the code churn would be larger.

vdeenadh replied:
Currently, there is an `Available` condition per PVC, which is false during `progressing` or `error` states. When there's a PV upload error, we could set the corresponding PVC condition to error with a msg indicating the reason. This PVC error condition gets folded up into global VRG condition as error. This is also in line with Raghu suggests: use only one condition per PVC.

Thus the VRG (global) condition will be in error either due to lack of PV backend data protection or PV cluster data protection. 

DRPC should allow a relocate action only if it sees that the VRG (global) condition is available. DRPC should proceed with the failover action requested by the user even if the VRG is not in available condition.

ShyamsundarR replied:
Ok, so we are modelling this as per-PVC and roll-up to a VRG condition. That works, thanks.

ShyamsundarR reviewed (commented):
Initial comments mostly to do with renaming.

@vdeenadh it would have been better to rename this a little later, as this changes the stack in the head around code and its order, but as you have started it let's possibly get some of it done.

I still need to review the changing order of annotation addition in the code, as that is critical.

vdeenadh commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -303,9 +303,9 @@ type VRGInstance struct {
-	pvcVRFinalizerProtected = "volumereplicationgroups.ramendr.openshift.io/pvc-vr-protection"
-	pvcInUse                = "kubernetes.io/pvc-protection"
+	vrgFinalizerName          = "volumereplicationgroups.ramendr.openshift.io/vrg-protection"
+	vrgProtectionPVCFinalizer = "volumereplicationgroups.ramendr.openshift.io/pvc-vr-protection"
```
Good point. 

While at it, I think we should rename:
pvcVRAnnotationProtectedKey to vrgProtectionPVCAnnotationKey ("vrg-protection")
pvcVRAnnotationProtectedValue to vrgProtectionPVCAnnotationValue ("protected")

Likewise for pvVRAnnotationRetentionKey.

ShyamsundarR replied:
> Good point.
> 
> While at it, I think we should rename:
> pvcVRAnnotationProtectedKey to vrgProtectionPVCAnnotationKey ("vrg-protection")

Suggest `pvcVRGProtectedAnnotationKey`

> pvcVRAnnotationProtectedValue to vrgProtectionPVCAnnotationValue ("protected")
> 
> Likewise for pvVRAnnotationRetentionKey.

vdeenadh replied:
My thought was along these lines: it is a PVCAnnotation (noun) and the purpose of the PVCAnnotation is to indicate VRGProtection (adjective). I was going by the common style of adjective-noun form of variable naming.

I guess you're see it as a VRGProtectionAnnotation applied to PVC, hence added a pvc prefix.

But, I'm fine with it if you want to keep the existing style. 

Should the key be "vrg-protection" and value be "configured"? The "configured" term does not say anything about the current state of protection, whereas, "protected" could lead to misinterpretation about the current state of protection.

ShyamsundarR replied:
Let's use what you suggested [here](https://github.com/RamenDR/ramen/pull/217#discussion_r691741490)

vdeenadh commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -885,13 +900,12 @@ func (v *VRGInstance) reconcileVRAsSecondary(pvc *corev1.PersistentVolumeClaim,
 
 	pvcNamespacedName := types.NamespacedName{Name: pvc.Name, Namespace: pvc.Namespace}
-	if err := v.processVRAsSecondary(pvcNamespacedName, log); err != nil {
+	if err := v.protectPVBackendDataAsSecondary(pvcNamespacedName, log); err != nil {
```
I merely mirrored the change for this twin.  Yes, I agree they both should match.  I will resolve this depending on what we conclude about the other one.

vdeenadh commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -773,16 +775,29 @@ func (v *VRGInstance) reconcileVRsAsPrimary() bool {
 
-		log.Info("Successfully processed VolumeReplication for PersistentVolumeClaim")
+		// Annotate that PVC protection is complete
+		if err := v.addProtectedAnnotationToPVC(pvc, log); err != nil {
```
Please do! 

I believe you may have originally added a "protecting" annotation and then switched it to "protected" annotation, about which I recall asking for some clarification at that time. Do you recall why you dropped the "protecting" annotation?  Did you drop it because the VRG finalizer on the PVC is doing double duty as the "protecting" annotation?

ShyamsundarR replied:
Moving the annotation here does not help us optimize the reconcile (which is what the annotation was meant for). As we would check for the annotation and not process steps in `prepareToProtectEligiblePVC` and that continues with moving the annotation here. As a result, it is possibly better left at the original location in the call stack of `prepareToProtectEligiblePVC`.

Each time we reconcile, we will be backing up the PV to the S3 store post this change for all PVCs that are part of the selector. If we are to use the annotation to skip backing up the PV to the S3 store, then I am not sure if we can detect and handle resize requests, hence optimizing this step may not be needed (and would possibly solve the resize issue as well).

vdeenadh commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -773,16 +775,29 @@ func (v *VRGInstance) reconcileVRsAsPrimary() bool {
+		}
 
+		// Protect the PV cluster data (in etcd)
+		if err := v.protectPVClusterData(pvc, log); err != nil {
```
This method is intended to protect PV Cluster Data (PV resource in the initial release, but, in future, this method will also protect PV snapshot resources).  Specifically, it is not protecting "PVC" cluster data. I agree it will be consistent with other naming, but I was worried that using the term PVC may be misleading that Ramen is protecting PVC cluster data, isn't it?

ShyamsundarR replied:
The confusion here is are we abstracting the methods as prepare PVC -> protect PV data -> protect PV cluster data, because of the renaming of `processVRAsPrimary` as `protectPVBackendDataAsPrimary` as we are dropping VR there (the change from process to protect is fine btw).

If the above were renamed from `processVRAsPrimary` to `protectVRAsPrimary`, it automatically qualifies as protecting backend data and acting on a VR. Then this method can be named `protectPVClusterData` without the PVC name. The flow is prepare PVC -> protect VR -> protect PV cluster data hence.

vdeenadh replied:
> protect VR
The method is `not` protecting the VR though. The method is actually protecting the PV (this is what it does) by creating a VR (this is how it does it).

vdeenadh replied:
If you want to retain the VR in the name at this level, then I'll either leave the method name as `processVRAsPrimary` or `createVRAsPrimary`.

ShyamsundarR replied:
`proess` then, not create as it may create or update

vdeenadh commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -658,7 +658,7 @@ func (v *VRGInstance) reconcileVRForDeletion(pvc *corev1.PersistentVolumeClaim,
 			return !requeue
 		}
-	} else if err := v.processVRAsPrimary(pvcNamespacedName, log); err != nil {
+	} else if err := v.protectPVBackendDataAsPrimary(pvcNamespacedName, log); err != nil {
```
I was trying to clarify that the act of protecting the PVC involves two major steps: protecting the PV Cluster Data (or PV resource now, but also PV snapshot resources in future) and protecting the PV backend storage volume, which is why I thought this method requires the qualifier "PVBackendData" to distinguish it from "PVClusterData".  I'm happy to consider different names for both of these, but I think it is important to distinguish it somehow to bring clarity to the two important VRG functions.

vdeenadh commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -607,17 +607,17 @@ func (v *VRGInstance) processForDeletion() (ctrl.Result, error) {
 		log := v.log.WithValues("pvc", pvcNamespacedName.String())
 
-		requeueResult, skip := v.preparePVCForVRProtection(pvc, log)
+		requeueResult, skip := v.prepareToProtectEligiblePVC(pvc, log)
```
> I think this function renaming should change, to drop eligible from the name. It's return values suggest appropriate eligibility actions for now.

Just for context, this method was originally named `preparePVCForVRProtection()` that in turn invoked the `protectPVC()` method.

Let me try to explain how I ended up adding the `Eligible` qualifier. Protection of PVC involves two main tasks: enabling replication via VolRep (VR) and uploading the PV resource to S3 (and in future, also uploading the snapshot resources, etc).  Given the main change I was making in the PR (moving the S3 upload and moving the protected annotation to a later point in the flow), I thought that the name `protectPVC()` no longer is the right fit for this method and can be even misleading, as it will be merely adding a finalizer and retaining the PV.  In my mind, the addition of the finalizer is a preparation not only required to clean up the VR but also to clean up S3. Hence, I thought this method is better named `prepareToProtectPVC()`.  I followed the existing imperative style of naming the method `protectPVC()` and renamed it to `prepareToProtectPVC()`. Given this, I had to differentiate the caller `preparePVCForVRProtection()` to `prepareToProtectEligiblePVC()`, which was doing eligibility checks. 

> The function specifically prepares a PVC to be replicated by a VR, as some of the conditions within the function are VR specific and not generic.

Please correct me if I'm mistaken, but with the change I'm making, I think the preparation phase is no longer limited to VR (which is why I got rid of the VR qualifier). To me, the preparation step of adding a finalizer is required to clean up S3 data.  I believe the preparation step is being used as a two phase commit operation: finalizer at the start and annotation at the end.  If the finalizer is present, it indicates that we have started the series of protection steps, like s3 upload, VR creation, etc., but without the presence of the annotation, we don't know which of those steps have completed successfully. During deletion of VRG, if the finalizer is present on PVC we need to clean up S3,  delete the VR, etc. Hence, this step is no longer VR specific in my mind.
 
In its present state, preparing the PVC involves checking if the PVC is eligible [done by `prepareToProtectEligiblePVC()`] which in turn does the preparation (initial steps of protecting the PVC) by adding a finalizer and retaining the PV [done by calling `prepareToProtectPVC()`]. I recall that there was some pending decision on whether PV retention is required or not. If we decide in future to not retain the PV, then `prepareToProtectPVC()` will be just adding the finalizer.

> Should it be renamed here, or should this call a generic preparePVCForReplication and that calls a protectPVCForVR is something we can do later? I would suggest we rename this as preparePVCForReplication

In my mind, the term "replication" is about protecting the PV backend data via VR and doesn't include backing up the PV cluster data (or resource).  I don't know if that's also how you use the term replication here. Whereas, as I outlined above, I believe this method is preparing the PVC for both: replication of PV backend data and backup of PV cluster data; hence, protection seems more appropriate, whereas replication seems more narrow.  

Thanks for the openness you all show by giving the renaming commit a fair look.  I understand that there's no good time to change names, because all those who are familiar with the code have to unlearn what they know.  Given that the change I was making (moving the S3 upload and moving the protected annotation to a later point in the flow) was disruptive enough, there is bound to be some relearning that is required.  Hence, I used it as an opportunity to look at what a method was being done to come up with descriptive names that match what it is doing today (not what it was doing in the past).  Hopefully, the new names reinforce the changed world better.  But, I'm happy to consider different names or keep the old names if that seems to be a better fit for what the method is doing today.

It would be easier if we all could read what's on each other's mind, but until then, we have to explain with long notes! :-)

ShyamsundarR replied:
I think abstracting this out to non-VR at present is too early, I would hence suggest we do something like `preparePVCForProtection` so that subsequent methods are protectPVCAsPrimary|Secondary, IOW prepare and then protect.

With the prepare -> protect -> preserve (for S3 backups) forming the 2 action steps?

ShyamsundarR replied:
I can see the following method rename here:
`preparePVCForVRProtection` -> `preparePVCForProtection`
`protectPVC` -> `prepareEligiblePVCForProtection`

vdeenadh replied:
I can't help but wonder about how the new names have reversed and yet makes sense. The power of diverse thought! I like it; will use these two.

Regarding protect vs preserve... VRG is in the job of protecting PVCs, which consists of two equally important parts:
1. Protecting the PV Contents: PV backend volume for now, PV snapshot contents in future (this is `what` it does). VRG creates VR CRs to enable replication of the storage volume (this is `how` it protects PV Contents).
2. Protecting the PV Cluster Data, aka PV related k8s objects or resources: PV for now, and snapshots in future (this is `what` it does). VRG backs up the PV Cluster Data to S3 (this is `how` it protects PV Cluster state).

If either of the above two fails, then the PV is not protected and the VRG state is not suitable for relocation.

ShyamsundarR replied:
> Regarding protect vs preserve... VRG is in the job of protecting PVCs, which consists of two equally important parts:
> 
>     1. Protecting the PV Contents: PV backend volume for now, PV snapshot contents in future (this is `what` it does). VRG creates VR CRs to enable replication of the storage volume (this is `how` it protects PV Contents).
> 
>     2. Protecting the PV Cluster Data, aka PV related k8s objects or resources: PV for now, and snapshots in future (this is `what` it does). VRG backs up the PV Cluster Data to S3 (this is `how` it protects PV Cluster state).
> 
> 
> If either of the above two fails, then the PV is not protected and the VRG state is not suitable for relocation.

I agree on the "what" and "how", all I am saying at present is that both of the "how" can change (protect PV cluster data may change to something else (replication cookies?) and instead of using VR we use other replication methods) and it is too early to confuse us with an abstraction that makes better sense in the future.

IOW I would like to reason with code as doing VR stuff and then doing PV stuff, than doing backend data stuff (which means VR) and doing cluster data (which means PV backups).

Hence the resistance to rename these as broadly as you suggest. We may/will end up here in a hopefully not too distant future, but let's wait till then.

vdeenadh replied:
I understand.

vdeenadh replied:
> We may/will end up here in a hopefully not too distant future, but let's wait till then.

I withdrew the renaming commit and kept the commits only to fixes.

ShyamsundarR commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -707,7 +720,7 @@ func (v *VRGInstance) reconcileVRsForDeletion() bool {
 		}
 
-		log.Info("Successfully processed VolumeReplication for PersistentVolumeClaim")
+		log.Info("Successfully protected PVC as secondary")
```
(nit) log message should not state "Secondary" Deletion may have happened as either primary or secondary. Do we really need to change the older message?

ShyamsundarR commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -861,13 +874,25 @@ func (v *VRGInstance) reconcileVRsAsPrimary() bool {
 		}
 
-		log.Info("Successfully processed VolumeReplication for PersistentVolumeClaim")
+		log.Info("Successfully protected the PVC as primary")
```
We setup logging with values state (key) and primary|secondary (value) and also filalize (key) with true (value) in processVRGActions, which helps us avoid primary|secondary words in the log message. I would hence stick to the older message here.

ShyamsundarR reviewed (commented):
(question) So we still retain the older single condition with this PR and split it into 2 conditions "Available" and "YetToBeNamed" subsequently?

Otherwise PR looks fine, we can address the log changes if the nits seem valid.
